photon-ml/src/main/scala/com/linkedin/photon/ml/event/EventEmitter.scala

```diff
@@ -0,0 +1,44 @@
+/*
+ * Copyright 2016 LinkedIn Corp. All rights reserved.
+ * Licensed under the Apache License, Version 2.0 (the "License"); you may
+ * not use this file except in compliance with the License. You may obtain a
+ * copy of the License at
+ *
+ * http://www.apache.org/licenses/LICENSE-2.0
+ *
+ * Unless required by applicable law or agreed to in writing, software
+ * distributed under the License is distributed on an "AS IS" BASIS, WITHOUT
+ * WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied. See the
+ * License for the specific language governing permissions and limitations
+ * under the License.
+ */
+package com.linkedin.photon.ml.event
+
+/**
+ * Base trait for all [[Event]] producers.
+ */
+trait EventEmitter {
+  private var registeredListeners = List[EventListener]()
+
+  /**
+   * Register an [[EventListener]] to consume [[Event]] objects produced by this class.
+   *
+   * @param listener The listener.
+   */
+  protected def registerListener(listener: EventListener): Unit = registeredListeners = listener :: registeredListeners
+
+  /**
+   * Shutdown and remove all currently registered listeners.
+   */
+  protected def clearListeners(): Unit = {
+    registeredListeners.foreach(_.close())
+    registeredListeners = List()
+  }
+
+  /**
+   * Produce an event for all registered listeners to consume.
+   *
+   * @param event The event.
+   */
+  protected def sendEvent(event: Event): Unit = registeredListeners.foreach(_.handle(event))
```

Comment:
We should call `_.handle` in a `try` block, so that we still notify subsequent listeners if a previous listener fails. Also, we probably shouldn't allow it to propagate up and kill the calling function.